### PR TITLE
Added BitBucket Server and Data Center support

### PR DESCRIPTION
**Pull request by lukefx**
Hi all, this is an initial implementation that works on BitBucket Server and Data Center.

Just enable it on configuration.toml and
```toml
git_provider="bitbucket_server"
```

And:
python cli.py --pr_url https://git.onpreminstanceofbitbucket.com/projects/PROJECT/repos/REPO/pull-requests/1 review

### CODE REVIEW

File: pr_agent/git_providers/bitbucket_server_provider.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+            "title": pr_title
+        })
+
+        response = requests.put(url=self.bitbucket_pull_request_api_url, headers=self.headers, data=payload)
+        return response
+
+    # bitbucket does not support labels
+    def publish_labels(self, pr_types: list):
+        pass
+
```
[the remark this answers is not in the document]
Comment (lukefx) [reply pr_agent/git_providers/bitbucket_server_provider.py:347]: Wondering if we should change that method and raise an exception when labels are not supported by bitbucket. To me, this sounds better as a pass.
Another way could be to create and raise a NotSupportedException.

**Comment by lukefx** [comment]
@mrT23 labels as comment is ugly as hell :) and you cannot search for them... I don't really think that this is useful in any way. Don't know if you ever seen that but bitbucket's server UI is pretty behind and old the current cloud version.

**Comment by mrT23** [comment]
> @mrT23 labels as comment is ugly as hell :) and you cannot search for them... This is not useful in any way. I don't know if you've seen that but Bitbucket's server UI is pretty behind and old with the current cloud version.

not a separate comment. for example, if you use custom labels, in 'describe', you will have an additional section in the PR header: 'PR custom Lables'.

**Comment by lukefx** [comment]
ok I will check and modify the code :)

**Comment by mrT23** [comment]
> ok I will check and modify the code :)

i don't think you need to modify for that anything. you do have the 'is_supported' method.

my main request is to add some documentation, otherwise it will be very hard to other people to use this mode

**Comment by vaibhavkumar-sf** [comment]
@CodiumAI-Agent /review

**Comment by vaibhavkumar-sf** [comment]
Preparing review...

**Comment by vaibhavkumar-sf** [comment]
Preparing review...

File: pr_agent/git_providers/bitbucket_server_provider.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+from starlette_context import context
+
+from .git_provider import FilePatchInfo, GitProvider, EDIT_TYPE
+from ..algo.pr_processing import find_line_number_of_relevant_line_in_file
+from ..algo.utils import load_large_diff
+from ..config_loader import get_settings
+from ..log import get_logger
+
+
+class BitbucketServerProvider(GitProvider):
```
Comment (vaibhavkumar-sf) [inline pr_agent/git_providers/bitbucket_server_provider.py:18]: **Suggestion:** Add a docstring to the `BitbucketServerProvider` class.
```suggestion
class BitbucketServerProvider(GitProvider):
    """
    Provides Git integration with Bitbucket Server.
    """
```

File: pr_agent/git_providers/bitbucket_server_provider.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+                   f"{self.pr.destination_branch}/.pr_agent.toml")
+            response = requests.request("GET", url, headers=self.headers)
+            if response.status_code == 404:  # not found
+                return ""
+            contents = response.text.encode('utf-8')
+            return contents
+        except Exception:
+            return ""
+
+    def publish_code_suggestions(self, code_suggestions: list) -> bool:
```
Comment (vaibhavkumar-sf) [inline pr_agent/git_providers/bitbucket_server_provider.py:64]: **Suggestion:** Add type hints to the function parameters of `BitbucketServerProvider.publish_code_suggestions()`.
```suggestion
    def publish_code_suggestions(self, code_suggestions: list[dict]) -> bool:
```

Comment (vaibhavkumar-sf) [inline pr_agent/git_providers/bitbucket_server_provider.py:64]: **Suggestion:** Add a docstring to the `publish_code_suggestions()` method.
```suggestion
    def publish_code_suggestions(self, code_suggestions: list[dict]) -> bool:
        """
        Publishes code suggestions as comments on the PR.
        """
```

File: pr_agent/git_providers/bitbucket_server_provider.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+            for comment in self.temp_comments:
+                self.remove_comment(comment)
+        except ValueError as e:
+            get_logger().exception(f"Failed to remove temp comments, error: {e}")
+
+    def remove_comment(self, comment):
+        pass
+
+    # funtion to create_inline_comment
+    def create_inline_comment(self, body: str, relevant_file: str, relevant_line_in_file: str):
```
Comment (vaibhavkumar-sf) [inline pr_agent/git_providers/bitbucket_server_provider.py:215]: **Suggestion:** Add type hints to the function parameters of `BitbucketServerProvider.create_inline_comment()`.
```suggestion
    def create_inline_comment(
      self, body: str, relevant_file: str, relevant_line_in_file: str
    ) -> dict:
```

Comment (vaibhavkumar-sf) [inline pr_agent/git_providers/bitbucket_server_provider.py:215]: **Suggestion:** Add a docstring to the `create_inline_comment()` method.
```suggestion
    def create_inline_comment(
      self, body: str, relevant_file: str, relevant_line_in_file: str
    ) -> dict:
        """
        Creates an inline comment for a code suggestion.
        """
```

File: pr_agent/git_providers/bitbucket_server_provider.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+        if position == -1:
+            if get_settings().config.verbosity_level >= 2:
+                get_logger().info(f"Could not find position for {relevant_file} {relevant_line_in_file}")
+            subject_type = "FILE"
+        else:
+            subject_type = "LINE"
+        path = relevant_file.strip()
+        return dict(body=body, path=path, position=absolute_position) if subject_type == "LINE" else {}
+
+    def publish_inline_comment(self, comment: str, from_line: int, file: str):
```
Comment (vaibhavkumar-sf) [inline pr_agent/git_providers/bitbucket_server_provider.py:230]: **Suggestion:** Add type hints to the function parameters of `BitbucketServerProvider.publish_inline_comment()`.
```suggestion
    def publish_inline_comment(
      self, comment: str, from_line: int, file: str
    ) -> requests.Response:
```

Comment (vaibhavkumar-sf) [inline pr_agent/git_providers/bitbucket_server_provider.py:230]: **Suggestion:** Add a docstring to the `publish_inline_comment()` method.
```suggestion
    def publish_inline_comment(
      self, comment: str, from_line: int, file: str
    ) -> requests.Response:
        """
        Publishes an inline comment on a specific line of a file in the PR.
        """
```

**Comment by vaibhavkumar-sf** [comment]
Preparing review...

**Comment by vaibhavkumar-sf** [comment]
Preparing review...

**Comment by mrT23** [comment]
@lukefx can you add a few lines fo documentation, so we can merge this ?

**Comment by lukefx** [comment]
> @lukefx can you add a few lines fo documentation, so we can merge this ?

Yes, I'm on it... I also added the support for webhooks and I will commit it soon. Sorry for the wait

**Comment by lukefx** [comment]
@mrT23 Please review the doc if is enough...

**Comment by mrT23** [comment]
@lukefx 
looks good. thanks for the contribution 👍

**Review by mrT23 — APPROVED** [review]

**Pull request merged**